service/common/src/main/java/org/apache/polaris/service/events/aws/cloudwatch/AwsCloudWatchEventListener.java

```diff
@@ -0,0 +1,275 @@
+        LOGGER.error(
+            "Error writing logs to CloudWatch - client-side error. Please adjust Polaris configurations: {}",
+            e.getMessage());
+      } else {
+        LOGGER.error("Error writing logs to CloudWatch - server-side error: {}", e.getMessage());
+      }
+      throw e;
+    } finally {
+      try {
+        reapFuturesMap();
+      } catch (Exception e) {
+        LOGGER.debug("Futures map could not be reaped: {}", e.getMessage());
+      }
+    }
+  }
+
+  private void sendToCloudWatch(InputLogEvent event) {
+    PutLogEventsRequest.Builder requestBuilder =
+        PutLogEventsRequest.builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .logEvents(List.of(event));
+
+    synchronized (this) {
+      if (sequenceToken != null) {
+        requestBuilder.sequenceToken(sequenceToken);
+      }
+
+      try {
+        executePutLogEvents(requestBuilder);
+      } catch (InvalidSequenceTokenException e) {
+        sequenceToken = getSequenceToken();
+        requestBuilder.sequenceToken(sequenceToken);
+        executePutLogEvents(requestBuilder);
+      }
+    }
+  }
+
+  private void executePutLogEvents(PutLogEventsRequest.Builder requestBuilder) {
+    PutLogEventsResponse response = client.putLogEvents(requestBuilder.build());
+    sequenceToken = response.nextSequenceToken();
+  }
+
+  private void reapFuturesMap() {
+    for (Future<?> future : futures.keySet()) {
+      if (future.isDone()) {
+        Future.State currFutureState = future.state();
+        EventWithRetry currValue = futures.remove(future);
+        if (currFutureState.equals(Future.State.FAILED)
+            || currFutureState.equals(Future.State.CANCELLED)) {
+          if (currValue.retryCount >= 3) {
+            LOGGER.error("Event retries failed. Event dropped: {}", currValue.inputLogEvent);
+          } else {
+            EventWithRetry newValue =
+                new EventWithRetry(currValue.inputLogEvent, currValue.retryCount + 1);
+            future =
+                executorService.submit(() -> sendAndHandleCloudWatchCall(newValue.inputLogEvent));
+            futures.put(future, newValue);
+          }
+        }
+      }
+    }
+  }
+
+  private void transformAndSendEvent(HashMap<String, Object> properties) {
+    properties.put("realm", callContext.getRealmContext().getRealmIdentifier());
+    properties.put("principal", securityContext.getUserPrincipal().getName());
+    // TODO: Add request ID when it is available
+    String eventAsJson;
+    try {
+      eventAsJson = objectMapper.writeValueAsString(properties);
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Error processing event into JSON string: {}", e.getMessage());
+      return;
+    }
+    InputLogEvent inputLogEvent = createLogEvent(eventAsJson, getCurrentTimestamp(callContext));
+    Future<?> future = executorService.submit(() -> sendAndHandleCloudWatchCall(inputLogEvent));
```

Comment:
Two thoughts:
1. For this listener, I think we should offer a simple synchronous submission mode
2. As we add more listeners, seems like some of this asynchronous handling logic will be duplicated so it might be worth refactoring it out. This can wait until we add more listeners though.